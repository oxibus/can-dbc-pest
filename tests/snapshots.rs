#![cfg(feature = "encoding")]

use std::fs;
use std::path::Path;

use can_dbc_pest::{decode_cp1252, DbcParser, Rule};
use insta::{assert_debug_snapshot, with_settings};
use pest::Parser;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [("tests/shared-test-files/dbc-cantools", "dbc-cantools")];
    for (path, file_name) in &test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(path);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: impl AsRef<Path>) {
    let path = path.as_ref();
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    if let Some(buffer) = decode_cp1252(&buffer) {
        match DbcParser::parse(Rule::file, &buffer) {
            Ok(pairs) => assert_debug_snapshot!(file_name, pairs),
            Err(e) => panic!("Failed to parse {file_name}.dbc: {e:#?}"),
        }
    } else {
        panic!("Failed to decode {file_name}.dbc as cp1252");
    }
}
