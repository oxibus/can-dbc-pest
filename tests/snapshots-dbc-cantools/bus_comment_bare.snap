---
source: tests/snapshots.rs
---
[
    Pair {
        rule: file,
        span: Span {
            str: "VERSION \"\"\n\n\nNS_ :\n\tNS_DESC_\n\tCM_\n\tBA_DEF_\n\tBA_\n\tVAL_\n\tCAT_DEF_\n\tCAT_\n\tFILTER\n\tBA_DEF_DEF_\n\tEV_DATA_\n\tENVVAR_DATA_\n\tSGTYPE_\n\tSGTYPE_VAL_\n\tBA_DEF_SGTYPE_\n\tBA_SGTYPE_\n\tSIG_TYPE_REF_\n\tVAL_TABLE_\n\tSIG_GROUP_\n\tSIG_VALTYPE_\n\tSIGTYPE_VALTYPE_\n\tBO_TX_BU_\n\tBA_DEF_REL_\n\tBA_REL_\n\tBA_DEF_DEF_REL_\n\tBU_SG_REL_\n\tBU_EV_REL_\n\tBU_BO_REL_\n\tSG_MUL_VAL_\n\nBS_:\n\nBU_:\n\nCM_ \"Comment1\";\nCM_ \"Comment2\";\n",
            start: 0,
            end: 380,
        },
        inner: [
            Pair {
                rule: line,
                span: Span {
                    str: "VERSION \"\"",
                    start: 0,
                    end: 10,
                },
                inner: [
                    Pair {
                        rule: version,
                        span: Span {
                            str: "VERSION \"\"",
                            start: 0,
                            end: 10,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "\"\"",
                                    start: 8,
                                    end: 10,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 10,
                    end: 11,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 11,
                    end: 12,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 12,
                    end: 13,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "NS_ :",
                    start: 13,
                    end: 18,
                },
                inner: [
                    Pair {
                        rule: new_symbols,
                        span: Span {
                            str: "NS_ :",
                            start: 13,
                            end: 18,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 18,
                    end: 19,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "NS_DESC_\n",
                    start: 20,
                    end: 29,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "NS_DESC_",
                            start: 20,
                            end: 28,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "CM_\n",
                    start: 30,
                    end: 34,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "CM_",
                            start: 30,
                            end: 33,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_DEF_\n",
                    start: 35,
                    end: 43,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_DEF_",
                            start: 35,
                            end: 42,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_\n",
                    start: 44,
                    end: 48,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_",
                            start: 44,
                            end: 47,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "VAL_\n",
                    start: 49,
                    end: 54,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "VAL_",
                            start: 49,
                            end: 53,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "CAT_DEF_\n",
                    start: 55,
                    end: 64,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "CAT_DEF_",
                            start: 55,
                            end: 63,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "CAT_\n",
                    start: 65,
                    end: 70,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "CAT_",
                            start: 65,
                            end: 69,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "FILTER\n",
                    start: 71,
                    end: 78,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "FILTER",
                            start: 71,
                            end: 77,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_DEF_DEF_\n",
                    start: 79,
                    end: 91,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_DEF_DEF_",
                            start: 79,
                            end: 90,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "EV_DATA_\n",
                    start: 92,
                    end: 101,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "EV_DATA_",
                            start: 92,
                            end: 100,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "ENVVAR_DATA_\n",
                    start: 102,
                    end: 115,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "ENVVAR_DATA_",
                            start: 102,
                            end: 114,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SGTYPE_\n",
                    start: 116,
                    end: 124,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "SGTYPE_",
                            start: 116,
                            end: 123,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SGTYPE_VAL_\n",
                    start: 125,
                    end: 137,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "SGTYPE_VAL_",
                            start: 125,
                            end: 136,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_DEF_SGTYPE_\n",
                    start: 138,
                    end: 153,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_DEF_SGTYPE_",
                            start: 138,
                            end: 152,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_SGTYPE_\n",
                    start: 154,
                    end: 165,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_SGTYPE_",
                            start: 154,
                            end: 164,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SIG_TYPE_REF_\n",
                    start: 166,
                    end: 180,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "SIG_TYPE_REF_",
                            start: 166,
                            end: 179,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "VAL_TABLE_\n",
                    start: 181,
                    end: 192,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "VAL_TABLE_",
                            start: 181,
                            end: 191,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SIG_GROUP_\n",
                    start: 193,
                    end: 204,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "SIG_GROUP_",
                            start: 193,
                            end: 203,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SIG_VALTYPE_\n",
                    start: 205,
                    end: 218,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "SIG_VALTYPE_",
                            start: 205,
                            end: 217,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SIGTYPE_VALTYPE_\n",
                    start: 219,
                    end: 236,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "SIGTYPE_VALTYPE_",
                            start: 219,
                            end: 235,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BO_TX_BU_\n",
                    start: 237,
                    end: 247,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BO_TX_BU_",
                            start: 237,
                            end: 246,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_DEF_REL_\n",
                    start: 248,
                    end: 260,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_DEF_REL_",
                            start: 248,
                            end: 259,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_REL_\n",
                    start: 261,
                    end: 269,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_REL_",
                            start: 261,
                            end: 268,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BA_DEF_DEF_REL_\n",
                    start: 270,
                    end: 286,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BA_DEF_DEF_REL_",
                            start: 270,
                            end: 285,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BU_SG_REL_\n",
                    start: 287,
                    end: 298,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BU_SG_REL_",
                            start: 287,
                            end: 297,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BU_EV_REL_\n",
                    start: 299,
                    end: 310,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BU_EV_REL_",
                            start: 299,
                            end: 309,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BU_BO_REL_\n",
                    start: 311,
                    end: 322,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "BU_BO_REL_",
                            start: 311,
                            end: 321,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "SG_MUL_VAL_\n",
                    start: 323,
                    end: 335,
                },
                inner: [
                    Pair {
                        rule: extended_identifier,
                        span: Span {
                            str: "SG_MUL_VAL_",
                            start: 323,
                            end: 334,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 335,
                    end: 336,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BS_:",
                    start: 336,
                    end: 340,
                },
                inner: [
                    Pair {
                        rule: bit_timing,
                        span: Span {
                            str: "BS_:",
                            start: 336,
                            end: 340,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 340,
                    end: 341,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 341,
                    end: 342,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BU_:",
                    start: 342,
                    end: 346,
                },
                inner: [
                    Pair {
                        rule: nodes,
                        span: Span {
                            str: "BU_:",
                            start: 342,
                            end: 346,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 346,
                    end: 347,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 347,
                    end: 348,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "CM_ \"Comment1\";",
                    start: 348,
                    end: 363,
                },
                inner: [
                    Pair {
                        rule: comment,
                        span: Span {
                            str: "CM_ \"Comment1\";",
                            start: 348,
                            end: 363,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "\"Comment1\"",
                                    start: 352,
                                    end: 362,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 363,
                    end: 364,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "CM_ \"Comment2\";",
                    start: 364,
                    end: 379,
                },
                inner: [
                    Pair {
                        rule: comment,
                        span: Span {
                            str: "CM_ \"Comment2\";",
                            start: 364,
                            end: 379,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "\"Comment2\"",
                                    start: 368,
                                    end: 378,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 379,
                    end: 380,
                },
                inner: [],
            },
            Pair {
                rule: EOI,
                span: Span {
                    str: "",
                    start: 380,
                    end: 380,
                },
                inner: [],
            },
        ],
    },
]
